### PR TITLE
[fix] add abstract pool

### PR DESCRIPTION
**Pull request by mosquito**
Inspired by #173 from @krieghan.

### CODE REVIEW

File: aio_pika/pool.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import asyncio
+import logging
+from typing import AsyncContextManager, TypeVar, Coroutine, Callable, Any
```
Comment (alvassin) [inline aio_pika/pool.py]: wrong imports order

File: tests/test_pool.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import asyncio
+
+import pytest
+from aio_pika.pool import Pool
+from tests import AsyncTestCase
```
Comment (alvassin) [inline tests/test_pool.py]: Why not used fixtures?

File: aio_pika/pool.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+
+T = TypeVar("T")
+ItemType = Coroutine[Any, None, T]
+ConstructorType = Callable[[], ItemType]
```
Comment (leenr) [inline aio_pika/pool.py]: `Awaitable[ItemType]`?
> `item = await self.__constructor()`

File: aio_pika/pool.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    def __init__(self, pool: Pool):
+        self.pool = pool
+        self.item = None
+
+    async def __aenter__(self) -> T:
+        self.item = await self.pool._get()
+        return self.item
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        self.pool.put(self.item)
```
Comment (leenr) [inline aio_pika/pool.py]: `self.item = None`?

File: aio_pika/pool.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    __slots__ = (
+        'loop', '__max_size', '__items',
+        '__constructor', '__created', '__lock'
+    )
+
+    def __init__(self, constructor: ConstructorType, max_size: int = None,
+                 loop: asyncio.AbstractEventLoop = None):
+
+        self.loop = loop or asyncio.get_event_loop()
+        self.__max_size = max_size or -1
```
Comment (leenr) [inline aio_pika/pool.py]: It seems that `-1` is not supported by the code below.
If `self.__max_size == -1`, `self._is_overflow` will be always `True`, and `await self.__items.get()` will hang in the `_get`.

File: tests/test_pool.py (hunk 2 above)
Comment (mosquito) [reply tests/test_pool.py]: All another test suite uses Unittest

**Comment by krieghan** [comment]
I tested this on my application under a pretty heavy load test.  It works without issues, and out-performs the pool that I'd thrown together.

**Pull request merged**